### PR TITLE
[REVIEW] Small refactor of `print_differences`

### PR DESCRIPTION
**Pull request by codereport**
While "debugging" and trying to understand the `CUDF_TEST_EXPECT_COLUMNS_EQUAL` code I got down the "call stack" to find a function `print_differences` that is responsible for 2 things:

1. Printing the differences
2. Performing the `EXPECT_EQ` on the difference `thrust::device_vector`

For a function called `print_differences` I was very surprised to find the final `EXPECT_EQ` inside - especially when you combine that with the fact that the first line of the function is:
```cpp
if (differences.size() <= 0) { return; }
```
The options are to remain the function to something like `check_if_any_differences_and_print` but whenever you have to put `AND` or `OR` into a function name - that is a sign you should refactor.

This small clean up job pulls out the `EXPECT_EQ` and renames `print_differences` accordingly. This makes the code much more clear.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/tests/utilities/column_utilities.cu [hunk 1]
```diff
@@ -209,18 +204,19 @@ struct column_comparator_impl {
...
                                      ComparatorType(*d_lhs, *d_rhs));
 
-    // shrink back down
-    differences.resize(thrust::distance(differences.begin(), diff_iter));
-    print_differences(differences, lhs, rhs, print_all_differences, depth);
+    differences.resize(thrust::distance(differences.begin(), diff_iter));  // shrink back down
+
+    if (not differences.empty()) {
+      auto const msg = differences_message(differences, lhs, rhs, print_all_differences, depth);
+      EXPECT_EQ(differences.size(), size_t{0}) << msg;
```
Comment (karthikeyann) [inline cpp/tests/utilities/column_utilities.cu:218]: This `EXPECT_EQ` will always fail inside this `if` block.

File: cpp/tests/utilities/column_utilities.cu [hunk 2]
```diff
@@ -176,21 +173,19 @@ void print_differences(thrust::device_vector<int> const& differences,
...
   } else {
-    //  If there are differences, just display the first one
-    int index = differences[0];
+    int index = differences[0];  // only stringify first difference
 
     auto diff_lhs = cudf::detail::slice(lhs, index, index + 1);
     auto diff_rhs = cudf::detail::slice(rhs, index, index + 1);
 
     std::vector<std::string> h_left_strings  = to_strings(diff_lhs);
     std::vector<std::string> h_right_strings = to_strings(diff_rhs);
```
Comment (karthikeyann) [inline cpp/tests/utilities/column_utilities.cu:184]: These 2 vectors are not used in this block.

Comment (karthikeyann) [reply cpp/tests/utilities/column_utilities.cu:184]: Compiler might be issuing a warning for this.

**Review by karthikeyann — CHANGES_REQUESTED** [review]
minor changes.

File: cpp/tests/utilities/column_utilities.cu (hunk 1 above)
Comment (codereport) [reply cpp/tests/utilities/column_utilities.cu:218]: That was the behaviour before as well. Is there a Google Test `FAIL` macro that can be used instead?

Comment (rgsl888prabhu) [reply cpp/tests/utilities/column_utilities.cu:218]: `ASSERT_FALSE`

File: cpp/tests/utilities/column_utilities.cu [hunk 3]
```diff
@@ -308,9 +304,12 @@ struct column_comparator_impl<list_view, check_exact_equality> {
...
       });
 
-    // shrink back down
-    differences.resize(thrust::distance(differences.begin(), diff_iter));
-    print_differences(differences, lhs, rhs, print_all_differences, depth);
+    differences.resize(thrust::distance(differences.begin(), diff_iter));  // shrink back down
+
+    if (not differences.empty()) {
+      auto const msg = differences_message(differences, lhs, rhs, print_all_differences, depth);
+      EXPECT_EQ(differences.size(), size_t{0}) << msg;
```
Comment (rgsl888prabhu) [inline cpp/tests/utilities/column_utilities.cu:311]: Same as above

File: cpp/tests/utilities/column_utilities.cu (hunk 2 above)
Comment (codereport) [reply cpp/tests/utilities/column_utilities.cu]: Great catch. Removed :)

File: cpp/tests/utilities/column_utilities.cu (hunk 3 above)
Comment (codereport) [reply cpp/tests/utilities/column_utilities.cu:308]: @rgsl888prabhu The problem is that we wan't this to fail all the time. `ASSERT_FALSE` won't fail when it is false. I will look into if there is a macro for this.

Comment (codereport) [reply cpp/tests/utilities/column_utilities.cu]: Found it. `GTEST_FAIL`. Replaced :)

File: cpp/tests/utilities/column_utilities.cu (hunk 1 above)
Comment (codereport) [reply cpp/tests/utilities/column_utilities.cu]: Replaced with `GTEST_FAIL`

File: cpp/tests/utilities/column_utilities.cu [hunk 4]
```diff
@@ -308,9 +301,12 @@ struct column_comparator_impl<list_view, check_exact_equality> {
         return lhs_offsets[index] == rhs_offsets[index] ? false : true;
       });
 
-    // shrink back down
-    differences.resize(thrust::distance(differences.begin(), diff_iter));
-    print_differences(differences, lhs, rhs, print_all_differences, depth);
+    differences.resize(thrust::distance(differences.begin(), diff_iter));  // shrink back down
+
+    if (not differences.empty()) {
+      auto const msg = differences_message(differences, lhs, rhs, print_all_differences, depth);
```
Comment (harrism) [inline cpp/tests/utilities/column_utilities.cu:307]: Why even create the variable `msg`? Could stream the output directly into FAIL...

**Review by harrism — APPROVED** [review]
Looks good.

File: cpp/tests/utilities/column_utilities.cu (hunk 4 above)
Comment (codereport) [reply cpp/tests/utilities/column_utilities.cu]: Good point :) Updated

File: cpp/tests/utilities/column_utilities.cu (hunk 3 above)
Comment (karthikeyann) [reply cpp/tests/utilities/column_utilities.cu]: `FAIL()` can be used. At this point, EXPECT_EQ seems trivially readable.

**Review by karthikeyann — APPROVED** [review]
LGTM

**Pull request merged**